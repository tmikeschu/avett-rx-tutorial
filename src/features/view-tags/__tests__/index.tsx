// src/features/view-tags/__tests__/index.tsx

import * as React from "react";
import * as utils from "test";

import { Tag, useGetTagsQuery } from "api";

import ViewTags from "..";

jest.mock("api", () => ({
  useGetTagsQuery: jest.fn(),
}));
const queryMock = useGetTagsQuery as jest.Mock;

describe("<ViewTags />", () => {
  it("shows a loading spinner while fetching data", async () => {
    queryMock.mockReturnValueOnce({
      loading: true,
    });
    const { getByTestId, container } = utils.render(<ViewTags />);
    expect(getByTestId("loading")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("shows an empty state", async () => {
    queryMock.mockReturnValueOnce({
      loading: false,
      data: {
        allTags: { data: [] },
      },
    });
    const { getByText, container } = utils.render(<ViewTags />);
    expect(getByText(/don't have any tags/i)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("shows an error state", async () => {
    jest.spyOn(console, "error").mockImplementation();
    queryMock.mockReturnValueOnce({
      loading: false,
      error: new Error("unable to fetch"),
    });
    const { getByText, container } = utils.render(<ViewTags />);
    expect(getByText(/oh no!/i)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
    jest.restoreAllMocks();
  });

  it("shows data after loading", async () => {
    queryMock.mockReturnValueOnce({
      loading: false,
      data: {
        allTags: {
          data: [{ name: "test tag", _id: "1" } as Tag],
        },
      },
    });
    const { queryByTestId, getByText, getByRole } = utils.render(<ViewTags />);
    expect(queryByTestId("loading")).toBeNull();
    expect(getByText(/test tag/i)).toBeInTheDocument();
    expect(getByRole("list", { name: /tags list/i })).toMatchSnapshot();
  });
});
