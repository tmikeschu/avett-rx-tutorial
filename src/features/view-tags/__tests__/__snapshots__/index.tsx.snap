// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ViewTags /> shows a loading spinner while fetching data 1`] = `
<div>
  <div
    class="chakra-spinner css-en9w25"
    data-testid="loading"
  >
    <span
      class="css-f8n5zr"
    >
      Loading...
    </span>
  </div>
</div>
`;

exports[`<ViewTags /> shows an empty state 1`] = `
<div>
  <p
    class="chakra-text css-9p4anf"
  >
    Oh snap! We don't have any tags to show yet.
  </p>
</div>
`;

exports[`<ViewTags /> shows an error state 1`] = `
<div>
  <p
    class="chakra-text css-ucebhq"
  >
    Oh no! Something went wrong fetching that data.
  </p>
</div>
`;

exports[`<ViewTags /> shows data after loading 1`] = `
<ul
  aria-label="tags list"
  class="css-1ttwu97"
  role="list"
>
  <li
    class="css-1ntkcqz"
  >
    test tag
  </li>
</ul>
`;
